### ISSUE TITLE
Intermittent test failure in TestS3Offload

### ISSUE THREAD

**Issue by merlimat** [apache/pulsar#2942]
```
Error Message
arrays differ firstly at element [15]; expected value is <48> but was <49>. 
Stacktrace
java.lang.AssertionError: arrays differ firstly at element [15]; expected value is <48> but was <49>. 
	at org.testng.Assert.fail(Assert.java:96)
	at org.testng.Assert.assertEquals(Assert.java:195)
	at org.testng.Assert.assertEquals(Assert.java:179)
	at org.apache.pulsar.tests.integration.offload.TestS3Offload.testPublishOffloadAndConsumeViaThreshold(TestS3Offload.java:212)
```

Full error logs at https://gist.github.com/merlimat/cf7823d2a5fafae7661772f91e1710a7

**Comment by ivankelly** [comment]
https://builds.apache.org/job/pulsar_precommit_integrationtests/4684/

Looks like a message is lost.

**Comment by jiazhai** [comment]
Seems it happens a lot these days.

**Comment by ivankelly** [comment]
@jiazhai can you post some example runs where you've seen it happen? Did it happen locally, or only on jenkins?

**Comment by jiazhai** [comment]
https://gist.github.com/jiazhai/f758298f5a7c53081d2c80b3a575e4b7

**Comment by jiazhai** [comment]
@ivankelly above is an example in PR #2998 
https://builds.apache.org/job/pulsar_precommit_integrationtests/5056/testReport/junit/org.apache.pulsar.tests.integration.offload/TestS3Offload/tiered_storage_test_suite/

**Comment by ivankelly** [comment]
The gist is a local run I guess, given the hanzi. What class of machine was it run on?

**Comment by ivankelly** [comment]
The issue is due to the broker being rebooted, but I'm not sure if the client behaviour is problematic or not. It's nothing to do with offload in any case.

The sequence is:
1. Client writes a bunch of messages
2. Offload kicks some segments out to S3.
3. The test deletes a ledger and reboots the broker to ensure readback.
4. The client starts reading back the topic from the very start.

The problem happens at the reboot and readback phase. The broker reboots and starts serving reads. 
The client manages to read a good bit. 
```
12:52:21.726 [TestNG-method=testPublishOffloadAndConsumeViaCLI-1:org.apache.pulsar.tests.integration.offload.TestS3Offload@149] INFO  org.apache.pulsar.tests.integration.offload.TestS3Offload - IKDEBUG read back entry 999 msgid 17:999:-1
12:52:22.134 [pulsar-client-io-23-1:org.apache.pulsar.client.impl.ClientCnx@183] INFO  org.apache.pulsar.client.impl.ClientCnx - [id: 0xe4bec5be, L:/127.0.0.1:52048 ! R:localhost/127.0.0.1:46664] Disconnected
12:52:22.135 [pulsar-client-io-23-1:org.apache.pulsar.client.impl.ConnectionHandler@109] INFO  org.apache.pulsar.client.impl.ConnectionHandler - [persistent://s3-offload-test-cli-aojz/ns1/topic1] [my-sub] Closed connection [id: 0xe4bec5be, L:/127.0.0.1:52048 ! R:localhost/127.0.0.1:46664] -- Will try again in 0.1 s
12:52:22.236 [pulsar-timer-25-1:org.apache.pulsar.client.impl.ConnectionHandler@112] INFO  org.apache.pulsar.client.impl.ConnectionHandler - [persistent://s3-offload-test-cli-aojz/ns1/topic1] [my-sub] Reconnecting after timeout
12:52:22.243 [pulsar-client-io-23-1:org.apache.pulsar.client.impl.ClientCnx@720] ERROR org.apache.pulsar.client.impl.ClientCnx - [id: 0xbad8308a, L:/127.0.0.1:52042 - R:localhost/127.0.0.1:46664] Close connection because received internal-server error org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: io.netty.channel.AbstractChannel$AnnotatedConnectException: syscall:getsockopt(..) failed: Connection refused: pulsar-broker-0/172.18.0.6:6650
12:52:22.244 [pulsar-client-io-23-1:org.apache.pulsar.client.impl.BinaryProtoLookupService@151] WARN  org.apache.pulsar.client.impl.BinaryProtoLookupService - [persistent://s3-offload-test-cli-aojz/ns1/topic1] failed to send lookup request : org.apache.pulsar.client.api.PulsarClientException$LookupException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: io.netty.channel.AbstractChannel$AnnotatedConnectException: syscall:getsockopt(..) failed: Connection refused: pulsar-broker-0/172.18.0.6:6650
java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException$LookupException: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: io.netty.channel.AbstractChannel$AnnotatedConnectException: syscall:getsockopt(..) failed: Connection refused: pulsar-broker-0/172.18.0.6:6650
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:647) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) [?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977) [?:1.8.0_191]
	at org.apache.pulsar.client.impl.ClientCnx.handleLookupResponse(ClientCnx.java:380) [pulsar-client-2.3.0-SNAPSHOT.jar:2.3.0-SNAPSHOT]
	at org.apache.pulsar.common.api.PulsarDecoder.channelRead(PulsarDecoder.java:118) [pulsar-client-2.3.0-SNAPSHOT.jar:2.3.0-SNAPSHOT]
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [pulsar-client-2.3.0-SNAPSHOT.jar:2.3.0-SNAPSHOT]
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [pulsar-client-2.3.0-SNAPSHOT.jar:2.3.0-SNAPSHOT]
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [pulsar-client-2.3.0-SNAPSHOT.jar:2.3.0-SNAPSHOT]
	at 
```

Then the broker realizes it can't get register, so kills itself.
```
12:52:21.754 [main] ERROR org.apache.pulsar.broker.loadbalance.impl.ModularLoadManagerImpl - Broker znode - [/loadbalance/brokers/pulsar-broker-0:8080] is own by different zookeeper-ssession 72259437335805978 
12:52:21.755 [main] ERROR org.apache.pulsar.broker.loadbalance.impl.ModularLoadManagerImpl - Unable to create znode - [/loadbalance/brokers/pulsar-broker-0:8080] for load balance on zookeeper 
```

When the broker comes back the subscription is reset back to the start (because we hadn't acked), but the client continues as if its reading from where it left off.

@merlimat @sijie @jiazhai 
So a couple of questions:
- Do we expect that users of the consumer keep track to ensure message ids are monotonically increasing if they want guarantees on the topic?
- Why can the client read from a broker before the broker has registered?

All info in build: https://builds.apache.org/job/pulsar_precommit_integrationtests/5164/

**Comment by ivankelly** [comment]
Also for completeness, the znode conflict issue is happening because the testcontainers integ tests are restarting the broker by killing the container and starting it again. The arquillian tests were using supervisorctl stop, which kills the broker gracefully.

**Comment by sijie** [comment]
> Do we expect that users of the consumer keep track to ensure message ids are monotonically increasing if they want guarantees on the topic?

Yes. Users should do that. And Reader is preferred over Consumer in this case.

> Why can the client read from a broker before the broker has registered?

it is probably related to #2935 ?

**Comment by ivankelly** [comment]
>> Do we expect that users of the consumer keep track to ensure message ids are monotonically increasing if they want guarantees on the topic?

> Yes. Users should do that. And Reader is preferred over Consumer in this case.

This is surprising, though I guess we don't state otherwise anywhere. I guess immediately acknowledging would fix it, though the unload change fixes the problem anyhow.

> it is probably related to #2935 ?

I thought the same, but this code is in the branch I was testing.

**Comment by ivankelly** [comment]
Looking at #2935, it doesn't actually do anything to prevent the broker from serving traffic before it's ready, it just doesn't advertize it. In this case, the client doesn't need to do a lookup, because it already has an assignment for this topic. Not sure if this is problematic.

**Comment by sijie** [comment]
@merlimat to read the whole thread

**Comment by ivankelly** [comment]
@merlimat pinging again.

**Comment by merlimat** [comment]
>  doesn't actually do anything to prevent the broker from serving traffic before it's ready, it just doesn't advertize it. 

That is correct. Ideally the 6650 port should only be opened when all the components of the service are ready to accept requests. 

> Not sure if this is problematic. 

The fix for #2935 was to cover for the problem of using schema service before it was actually initialized. I think this might actually still be a problem when the lookup is bypassed since the broker service get started before the schema registry service.

**Comment by ivankelly** [comment]
Ok, lets create an issue for this at least. I'll close this one.
